src/Symfony/Contracts/Service/ServiceSubscriberTrait.php

```diff
@@ -42,8 +43,60 @@ public static function getSubscribedServices(): array
                 continue;
             }
 
-            if (self::class === $method->getDeclaringClass()->name && ($returnType = $method->getReturnType()) && !$returnType->isBuiltin()) {
-                $services[self::class.'::'.$method->name] = '?'.($returnType instanceof \ReflectionNamedType ? $returnType->getName() : $returnType);
+            if (self::class !== $method->getDeclaringClass()->name) {
+                continue;
+            }
+
+            if (!$returnType = $method->getReturnType()) {
+                continue;
+            }
+
+            if ($returnType->isBuiltin()) {
+                continue;
+            }
+
+            if (\PHP_VERSION_ID >= 80000) {
+                if (\count($method->getAttributes(SubscribedService::class))) {
+                    continue;
+                }
+
+                trigger_deprecation('symfony/service-contracts', '5.4', 'Using the "%s" trait in "%s" without subscribed service methods marked with the "%s" attribute is deprecated and will not be allowed in Symfony 6.0.', ServiceSubscriberTrait::class, self::class, SubscribedService::class);
+            }
+
+            $services[self::class.'::'.$method->name] = '?'.($returnType instanceof \ReflectionNamedType ? $returnType->getName() : $returnType);
+        }
+
+        if (\PHP_VERSION_ID > 80000) {
+            foreach ((new \ReflectionClass(self::class))->getMethods() as $method) {
+                if (self::class !== $method->getDeclaringClass()->name) {
+                    continue;
+                }
+
+                if (!$attribute = $method->getAttributes(SubscribedService::class)[0] ?? null) {
+                    continue;
+                }
+
+                if ($method->isStatic() || $method->isAbstract() || $method->isGenerator() || $method->isInternal() || $method->getNumberOfRequiredParameters()) {
+                    // todo, what exception to throw?
+                    throw new \RuntimeException(sprintf('Cannot use "%s" on method "%s::%s()" (can only be used on non-static, non-abstract methods with no parameters).', SubscribedService::class, self::class, $method->name));
+                }
+
+                if (!$returnType = $method->getReturnType()) {
+                    // todo, what exception to throw?
+                    throw new \RuntimeException(sprintf('Cannot use "%s" on methods without a return type in "%s::%s()".', SubscribedService::class, $method->name, self::class));
+                }
+
+                if (!$returnType instanceof \ReflectionNamedType) {
+                    // todo, what exception to throw?
+                    throw new \RuntimeException(sprintf('Cannot use "%s" on methods with a union return type in "%s::%s()".', SubscribedService::class, $method->name, self::class));
```

Comment:
Why ? the previous loop does not have this restriction.

Reply:
I treated this as a bug. If the return type is `Service1|Service2`, what should be used for the service locator?